imputation/ts_impute.go

```diff
@@ -0,0 +1,103 @@
+/*
+Package imputation implements a library for time series imputation.
+The dates in the time series are assumed to be either of the forms of (all have the same format):
+"YYYY-MM-DD" or
+"YYYY-MM" or
+"YYYY"
+
+The imputation library first finds the minimum gap of existing points and then adds new data points with values according to the selected method.
+*/
+package imputation
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"sort"
+	"time"
+)
+
+// TimeSeries stores a time series.  Dates are expressed in ISO 8061 format as illustrated in the constants below.
+type TimeSeries map[string]float64
+
+// Assume 3 formattig for the dates: daily, monthly, and yearly.
+const (
+	dayfmt   = "2006-01-02"
+	monthfmt = "2006-01"
+	yearfmt  = "2006"
+)
+
+// The diff function returns the minimum gap (will change to GCD) from a given set of data points
+// that all have the same format.
+// TODO(eftekhari-mhs): add OK, ERR return values and handle errors.
+func diff(keys []string, dateFormat string) (int, int, int) {
+	year, month, day := 0, 0, 0
+
+	duration := 1000 //A large number.
```

Comment:
Can doing using something like this https://yourbasic.org/golang/max-min-int-uint/